### PR TITLE
Expression handling

### PR DESCRIPTION
**Pull request by kerschke**
This PR allows to finally use task dependent expressions in the learner (see #767).

### CODE REVIEW

**Comment by larskotthoff** [comment]
Thanks Pascal.

- Please document prominently what parameters are available. This should appear not only in the documentation for `evaluateLearner`, but also `makeLearner`, `setHyperPars` and anywhere else where people can set hyperparameters.
- Why is the task itself in the dictionary?
- In the unit tests, instead of checking the values for randomForest (which are specified somewhere else), could you please use a hand-constructed expression? Otherwise this test will fail for no good reason if the values in randomForest are changed.

**Comment by kerschke** [comment]
>    Please document prominently what parameters are available. This should appear not only in the documentation for evaluateLearner, but also makeLearner, setHyperPars and anywhere else where people can set hyperparameters.
Ok, I'll do that :)

> Why is the task itself in the dictionary?

That's a decision we've made with @mllg and @berndbischl. The main reason is that now the user can use any expression which takes any information from the task. I mean, you probably wouldn't want to do `expression(task)` but `expression(task$...)`.

> In the unit tests, instead of checking the values for randomForest (which are specified somewhere else), could you please use a hand-constructed expression? Otherwise this test will fail for no good reason if the values in randomForest are changed.

The tests already contain hand-constructed expressions, e.g. the rpart with hand-constructed hyperparameters.. The reason for having the randomForest in there, is actually to check whether it also works with the parameter setup that is defined from our side within the definition of the learner. And if those values are changed by someone of us, well then that person should also update the unit tests accordingly. Right?

**Comment by larskotthoff** [comment]
> > Why is the task itself in the dictionary?  
> 
> That's a decision we've made with @mllg and @berndbischl. The main reason is
> that now the user can use any expression which takes any information from the
> task. I mean, you probably wouldn't want to do `expression(task)` but
> `expression(task$...)`.

Ok, but then this should be documented as well.

> > In the unit tests, instead of checking the values for randomForest (which
> > are specified somewhere else), could you please use a hand-constructed
> > expression? Otherwise this test will fail for no good reason if the values
> > in randomForest are changed.  
> 
> The tests already contain hand-constructed expressions, e.g. the rpart with
> hand-constructed hyperparameters.. The reason for having the randomForest in
> there, is actually to check whether it also works with the parameter setup
> that is defined from our side within the definition of the learner. And if
> those values are changed by someone of us, well then that person should also
> update the unit tests accordingly. Right?

I don't see the benefit of testing something external -- what does this test
that the hand-constructed example doesn't test? It introduces unnecessary
coupling and I'm sure that we'll always forget this and then have to come back
to a PR later and fix things.

**Comment by kerschke** [comment]
the documentation has been updated.. please review again and then merge ;)

**Comment by mllg** [comment]
Reviewed. Everything okay. @berndbischl -> ok to merge?

**Comment by kerschke** [comment]
@berndbischl: ping..

**Comment by berndbischl** [comment]
you do know that i am on holiday right?

**Comment by kerschke** [comment]
yes, and I see your working on all other ends of the package during your holiday ;)

**Comment by berndbischl** [comment]
evaluateParset should be called evaluateParamSet. We have to be consistent with the names

**Comment by kerschke** [comment]
Then, we're mixing up the names from ParamHelpers (evaluateParamSet) and mlr (evaluateParSet). But if that's the only thing that you don't like, I can fix that tomorrow..

**Comment by berndbischl** [comment]
actually, thinking about this, the whole name scheme seems not good as too general.
evaluateLearner can mean a million things. especially in the context of mlr. 
evaluateExpressions would be a lot better. probably this should be S3 then.

**Comment by berndbischl** [comment]
evaluateLearner.R: the arg "parset" should use a rox template. it adds no new documentation. its also wrong as "requires" can now contain both "quote" and "expression".

furthemore: evaluateParset and evaluateLearner do not have any arg checks with checkmate

**Comment by berndbischl** [comment]
> Learners can contain task dependent expressions, see \code{\link{evaluateLearner}} for more information.

it's not really the learner, but the paramset of the learner that can contain these expressions

**Comment by kerschke** [comment]
> evaluateLearner.R: the arg "parset" should use a rox template. it adds no new documentation. its also wrong as "requires" can now contain both "quote" and "expression".

so `mlr::evaluateLearner` should be `mlr::evaluateExpression.learner`, `mlr::evaluateParSet` should be `mlr::evaluateExpression.ParamSet`. And `ParamHelpers::evaluateParamSet` stays the way it is?

> furthemore: evaluateParset and evaluateLearner do not have any arg checks with checkmate

I avoided those checks, as it is only called from within `train` and `tuneParams` - and in both cases the arguments are checked prior to calling `evaluateLearner`.

**Comment by kerschke** [comment]
> it's not really the learner, but the paramset of the learner that can contain these expressions

The hyperparameters can contain expressions as well..

**Comment by berndbischl** [comment]
> The hyperparameters can contain expressions as well..

which is pretty much underdocumented. that must be precisely documented. saying "somewhere in the learner" does not help.

**Comment by kerschke** [comment]
Did you look at the examples?

**Comment by berndbischl** [comment]
> Did you look at the examples?

for what?

**Comment by kerschke** [comment]
Well, if you configure a learner, you most likely define the values via the hyper parameters. And there are examples given, e.g. `makeLearner("classif.rpart", minsplit = expression(k))`.
But if it's not obvious enough, I can re-iterate over it..

**Comment by berndbischl** [comment]
> Well, if you configure a learner, you most likely define the values via the hyper parameters. And there are examples given, e.g. makeLearner("classif.rpart", minsplit = expression(k)).
> But if it's not obvious enough, I can re-iterate over it..

a) i am not saying i dislike the feature of having expression in the par.vals of makeLearner and setHyperPars

b) of course i am also reading the examples

c) i REALLY do hope that you dont confuse API docs with examples. the former must be complete, while the latter should demonstrate interesting use cases. why do we have to discuss this?

**Comment by berndbischl** [comment]
> furthemore: evaluateParset and evaluateLearner do not have any arg checks with checkmate
> I avoided those checks, as it is only called from within train and tuneParams - and in both cases the arguments are checked prior to calling evaluateLearner.

you have exported the functions, please add the checks

**Comment by kerschke** [comment]
> a) i am not saying i dislike the feature of having expression in the par.vals of makeLearner and setHyperPars

Me neither. I just wanted to respond to your comment ("it's not really the learner, but the paramset of the learner that can contain these expressions") by clarifying that from my point of view it's actually the learner and not just it's paramset.

> b) of course i am also reading the examples

good :)

> c) i REALLY do hope that you dont confuse API docs with examples. the former must be complete, while the latter should demonstrate interesting use cases. why do we have to discuss this?

I am not mixing it up. But from my point of view, examples are useful to clarify exactly those things. But as I said before, I can re-iterate over it and add a sentence or two..

**Comment by berndbischl** [comment]
example comment: tuning with expressions seems to be one of the most important use cases, so we should add an example in tuneParams. must execute quickly.

**Comment by kerschke** [comment]
Ok, I might take the example that's already mentioned in the tutorial documentation..

**Comment by berndbischl** [comment]
> so mlr::evaluateLearner should be mlr::evaluateExpression.learner, mlr::evaluateParSet should be mlr::evaluateExpression.ParamSet. And ParamHelpers::evaluateParamSet stays the way it is?

hmm. this is hard and needs to be defined now. 
evaluateParamExpressions? 
and make this an S3 generic in PH.

signatures would then be

PH::evaluateParamExpressions(obj, ....)
PH::evaluateParamExpressions.ParamSet(obj, dict, ....)

mlr::evaluateParamExpressions.LearnerParamSet(obj, task, ....)
mlr::evaluateParamExpressions.Learner(obj, task, ....)

@larskotthoff  @mllg i guess i want some help here

the thing abobe with the generic over multiple packages and multiple classes that dont inherit from the same base class now violates what i said and wanted before. but to me it seems to be still the best fitting option....

**Comment by berndbischl** [comment]
> mlr::evaluateParamExpressions.LearnerParamSet(obj, task, ....)

fuck that does not work. i wanted to attach a more specific class LearnerParamSet to the par.set in makeRLearner. that would work.
but we also need this for the tuning param set.

**Comment by larskotthoff** [comment]
Hmm. I'm not a fan of this -- it couples PH and mlr more tightly and I think that this is something that should be avoided. Could we simply have a non-S3 function `evaluateParamExpression` that is passed a dictionary/environment? Then we could use this in other places as well (e.g. for resampling reduce number of folds for small numbers of data points).

**Comment by berndbischl** [comment]
ok new try:
PH: evaluateParamSetExpressions(par.set, dict)
this we already have. (must be simply renamed)

mlr:
we export makeTaskDisctionary. better call it getTaskDictionary
mlr::evaluateLearnerExpressions(lrn, task)
mlr::evaluateParamSetExpression with a task we then dont have. the user has to call
getTaskDictionary first, then evaluateParamSetExpressions
(we could later or now introduce evaluateLearnerParamSet(par.set, task)?)

**Comment by berndbischl** [comment]
or this is all really bad design and this is much simpler:
we ONLY have in PH:
evaluateParamExpressions.Par(obj, dict)
evaluateParamExpressions.ParamSet(obj, dict)
evaluateParamExpressions.list(obj, dict)

in mlr we have zero extra function, except for getTaskDictionary.
we tell the user anyhow that he does not have to call these new evaluateBla function. so we dont really need to save the one line with the getTaskDictionary. 

I think i like this best.

**Comment by larskotthoff** [comment]
Yep, I like you last suggestion best.

**Comment by berndbischl** [comment]
@kerschke 
give @mllg a chance to comment before you start changes

**Comment by mllg** [comment]
```{r}
evaluateParamExpressions.Par(obj, dict)
evaluateParamExpressions.ParamSet(obj, dict)
evaluateParamExpressions.list(obj, dict)
```
This is good. And we additionally define `evaluateParamExpressions.Learner(obj, dict)` in mlr? I don't see why this should not work.

**Comment by kerschke** [comment]
Lets summarize it:

1) Code changes to PH:
```{r}
evaluateParamExpressions.Par(obj, dict)
evaluateParamExpressions.ParamSet(obj, dict)
evaluateParamExpressions.list(obj, dict)
```

2) Code changes to mlr:
```{r}
evaluateParamExpressions.learner(obj, dict)
getTaskDictionary(task) # just renaming makeTaskDictionary

```

3) Documentation should be more explicit on where the expressions come from, i.e. do we evaluate a Param, ParamSet, Learner (containing ParamSet and HyperPars), etc.

4) Add a short example to `tuneParams`.

5) Update the tutorial accordingly..

Is that all or am I missing something?
By the way, where / when do we need to `evaluateParamExpressions.list(obj, dict)`?

**Comment by berndbischl** [comment]
> By the way, where / when do we need to evaluateParamExpressions.list(obj, dict)?

to eval the epressions in the par.vals list?

**Comment by berndbischl** [comment]
otherwise i think i agree with the summary list. go.

**Comment by kerschke** [comment]
>> By the way, where / when do we need to evaluateParamExpressions.list(obj, dict)?

> to eval the epressions in the par.vals list? 

Ok, that makes sense :)

> otherwise i think i agree with the summary list. go.

cool! then, i'll work on it and let you know, once i'm done. in the meantime, enjoy your holidays ;)

**Comment by kerschke** [comment]
>  i'll work on it and let you know, once i'm done.

Part 1 (updating PH) should be done. The [PR](https://github.com/berndbischl/ParamHelpers/pull/136) is open / ready for review. Let me know, if there are things that you don't like about it. Otherwise, I will go on and update mlr accordingly.

**Comment by kerschke** [comment]
Solved 2) and 4).
Let me know, whether there are still functions remaining, where the documentation should be more explicit.

Once everything's ok from your side (and ready to merge), I can update the tutorial accordingly.

**Comment by mllg** [comment]
Everything looks good. Will merge soon if Travis agrees.

**Comment by mllg** [comment]
@berndbischl @larskotthoff Last call. :wink:

@kerschke Where can I find your tutorial stuff? Can you link it here?

**Comment by berndbischl** [comment]
> @berndbischl @larskotthoff Last call

i am on holiday pls wait at least until monday.

**Comment by schiffner** [comment]
> Where can I find your tutorial stuff? Can you link it here?

In case that you haven't found it already: https://github.com/mlr-org/mlr-tutorial/pull/49

**Comment by larskotthoff** [comment]
Looks good to me.

**Comment by mllg** [comment]
@berndbischl ping ping ping

**Comment by kerschke** [comment]
@berndbischl what's your status on this one? i mean ping, ping, ping ;)

**Comment by jakob-r** [comment]
Test fail because of roxygen2 6.0.0

**Comment by jakob-r** [comment]
@berndbischl please review

**Comment by mllg** [comment]
@berndbischl ping

**Comment by jakob-r** [comment]
@berndbischl ping

File: R/train.R [hunk 1]
```diff
@@ -31,6 +31,10 @@
 train = function(learner, task, subset, weights = NULL) {
   learner = checkLearner(learner)
   assertClass(task, classes = "Task")
+  if (hasExpression(learner)) {
+    dict = getTaskDictionary(task = task)
```
Comment (mb706) [inline R/train.R]: The `dict` should somehow contain information about `subset`. The learner only sees the subset task, and probably expects the parameters to behave accordingly.

Comment (jakob-r) [inline R/train.R]: I would assume that the task is already subsetted?

Comment (mb706) [inline R/train.R]: Subsetting happens in `trainLearner` ([e.g.](https://github.com/mlr-org/mlr/blob/c2d40a3264ef564391260e38acf5be4d8c4f7473/R/RLearner_classif_logreg.R#L22))

Comment (jakob-r) [inline R/train.R]: Of course. I remember... what I don't remember is the motivation behind this though.

**Comment by jakob-r** [comment]
All reviews were postive. @mb706 comment can be implemented later in another PR.

**Comment by berndbischl** [comment]
@jakob-r @mllg 
can we please comment on why:
a) you merged something that apparently seems wrong? or please tell me why @mb706 is not correct.
if the the subsetting is done later, the the "n" symbol will we incorrect in the dict?

if thats the case, that creates invalid results. 

b) this is a base change. how where the 2 core members who reviewed positively?

**Comment by berndbischl** [comment]
@jakob-r 
i am sorry, i wanted to avoid this, but if it takes 3 min (!) to check EXACTLY what @mb706 said in a test.

```
load_all()
task = makeClassifTask(data = iris, target = "Species")
dict = getTaskDictionary(task = task)
lrn1 = makeLearner("classif.rpart", minsplit = expression(n))
print(getHyperPars(lrn1))
m = train(lrn1, task)
print(m$learner.model$control$minsplit)
m = train(lrn1, task, subset = 1:100)
print(m$learner.model$control$minsplit # is 150! wrong
```

what do you think happens now to people who use that code, and generate results from it?

will revert now

**Comment by jakob-r** [comment]
`n` is documented to be the size of the Task and not of the subset. 

R/evaluateParamExpressions.R
> \item{\code{n}:} the number of observations in the task

**Comment by berndbischl** [comment]
> In other words this example just assumes something that is never mentioned in the documentation.

we can discuss this on hangout, but you cannot be serious

**Comment by berndbischl** [comment]
maybe as reviewer it is also your obligation to think about whether the docs make sense? whether the semantics are good and proper? 

do you want to argue they are, here? that the example i posted is OK? you would like to have that behavior in your own experiments? that the defaults in the other underlying packages work this way?

**Pull request merged**